src/test/java/lambda/part3/exercise/Mapping.java

```diff
@@ -165,7 +163,189 @@ public LazyFlatMapHelper(List<T> list, Function<T, List<R>> function) {
         }
     }
 
+    interface ReachIterable<T> {
+        boolean tryGet(Consumer<T> consumer);
+
+        static<T> ReachIterable<T> from(List<T> list) {
+            final Iterator<T> iterator = list.iterator();
+            return new ReachIterable<T>() {
+                @Override
+                public boolean tryGet(Consumer<T> consumer) {
+                    if (iterator.hasNext()) {
+                        consumer.accept(iterator.next());
+                        return true;
+                    } else {
+                        return false;
+                    }
+                }
+            };
+        }
+
+        default List<T> force() {
+            final List<T> result = new ArrayList<>();
+            while (tryGet(result::add));
+            return result;
+        }
+
+        default ReachIterable<T> filter(Predicate<T> predicate) {
+            final ReachIterable<T> self = this;
+            return new ReachIterable<T>() {
+                @Override
+                public boolean tryGet(Consumer<T> consumer) {
+                    return self.tryGet(t -> {
+                        if (predicate.test(t)) {
+                            consumer.accept(t);
+                        }
+                    });
+                }
+            };
+        }
+
+        default <R> ReachIterable<R> map(final Function<T, R> function) {
+            final ReachIterable<T> self = this;
+            return new ReachIterable<R>() {
+                @Override
+                public boolean tryGet(final Consumer<R> consumer) {
+                    return self.tryGet(t -> consumer.accept(function.apply(t)));
+                }
+            };
+        }
+
+        default <R> ReachIterable<R> flatMap(Function<T, ReachIterable<R>> function) {
+            final ReachIterable<T> self = this;
+            return new ReachIterable<R>() {
+                private ReachIterable<R> reachIterable;
+                @Override
+                public boolean tryGet(final Consumer<R> consumer) {
+                    boolean res = self.tryGet(t -> reachIterable = function.apply(t));
+                    while (reachIterable.tryGet(consumer));
+                    return res;
+                }
+            };
+        }
+
+        default boolean anyMatch(Predicate<T> p) {
+            return firstMatch(p).isPresent();
+        }
+
+        default boolean allMatch(Predicate<T> predicate) {
+            return !anyMatch(predicate.negate());
+        }
+
+        default boolean noneMatch(Predicate<T> predicate) {
+            return !anyMatch(predicate);
+        }
+
+        default Optional<T> firstMatch(Predicate<T> predicate) {
+            final Object[] objects = new Object[1];
+            while (tryGet(t -> {

```

Comment:
Incorrect. `firstMatch` should be able to return value in case of infinite stream.